### ISSUE TITLE
Problems with bit.so

### ISSUE THREAD

**Issue by johnwcowan** [ashinn/chibi-scheme#406]
When I released the sample implementation of SRFI 151, I provided a code path specific to Chibi that makes use of bit.{so,dll}, but the test suite generates a lot of errors that do not appear on other systems tested (Chicken, Gauche, Larceny, Kawa, Sagittarius).  They do not appear in Chibi either when I remove the Chibi-specific code path and run the (slow) generic code from SLIB, so the `cond-expand` is currently commented out in `srfi-151.sld`.  Note that some of the tests are duplicates, because I merged several different SRFI-33 and SRFI-60 test suites that had some overlap. 
 Here are the errors:

```
bitwise: ...............xxxx..................................................
..............................................................................
...............................xxxxx......xxx................xxx............xx
....................................x...x......................x..............
.xxxx.........................................................................
..............................................................................
....xxxxx.............................................................x...x...
....................x
526 out of 558 (94.3%) tests passed in 0.0936200618743896 seconds.
32 failures (5.7%).
FAIL: (bitwise-ior 1694076839 -4290775858)
    expected -2600468497 but got 18446744071109083119
FAIL: (bitwise-ior -193073517 1689392892)
    expected -184549633 but got 4110417663
FAIL: (bitwise-xor 1694076839 -4290775858)
    expected -2604237975 but got 18446744071105313641
FAIL: (bitwise-xor -193073517 1689392892)
    expected -1865418641 but got 2429548655
FAIL: (copy-bit 30 -1 #t)
    expected -1 but got 4294967295
FAIL: (copy-bit 31 -1 #t)
    expected -1 but got 4294967295
FAIL: (copy-bit 62 -1 #t)
    expected -1 but got 4611686022722355199
FAIL: (copy-bit 63 -1 #t)
    expected -1 but got 9223372041149743103
FAIL: (copy-bit 128 -1 #t)
    expected -1 but got 340282366920938463463374607436063178751
FAIL: (copy-bit 64 340282366920938463463374607431768211456 #f)
    expected 340282366920938463463374607431768211456 but got 0
FAIL: (copy-bit 64 -340282366920938463463374607431768211456 #f)
    expected -340282366920938463463374607431768211456 but got 0
FAIL: (copy-bit 256 -340282366920938463463374607431768211456 #t)
    expected -340282366920938463463374607431768211456 but got 115792089237316195423570985010149409490260605216923310408710492420137293971456
FAIL: (bit-field-rotate 340282366920938463463374607431768211456 128 0...
    expected 340282366920938463463374607431768211456 but got 0
FAIL: (bit-field-rotate 340282366920938463463374607431768211457 3 0 64)
    expected 340282366920938463463374607431768211464 but got 8
FAIL: (bit-field-rotate 340282366920938463463374607431768211457 -3 0 64)
    expected 340282366920938463465680450440981905408 but got 2305843009213693952
FAIL: (bit-field-reverse -2 0 31)
    expected -1073741825 but got 18446744072635809791
FAIL: (bit-field-reverse -2 0 32)
    expected -2147483649 but got 18446744071562067967
FAIL: (arithmetic-shift -100000000000000000000000000000000 -100)
    expected -79 but got -78
FAIL: (arithmetic-shift -100000000000000000000000000000000 -100)
    expected -79 but got -78
FAIL: (first-set-bit (expt -2 99))
    expected 99 but got 60
FAIL: (bitwise-ior 1694076839 -4290775858)
    expected -2600468497 but got 18446744071109083119
FAIL: (bitwise-ior -193073517 1689392892)
    expected -184549633 but got 4110417663
FAIL: (bitwise-xor 1694076839 -4290775858)
    expected -2604237975 but got 18446744071105313641
FAIL: (bitwise-xor -193073517 1689392892)
    expected -1865418641 but got 2429548655
FAIL: (copy-bit 30 -1 #t)
    expected -1 but got 4294967295
FAIL: (copy-bit 31 -1 #t)
    expected -1 but got 4294967295
FAIL: (copy-bit 62 -1 #t)
    expected -1 but got 4611686022722355199
FAIL: (copy-bit 63 -1 #t)
    expected -1 but got 9223372041149743103
FAIL: (copy-bit 128 -1 #t)
    expected -1 but got 340282366920938463463374607436063178751
FAIL: (arithmetic-shift -100000000000000000000000000000000 -100)
    expected -79 but got -78
FAIL: (arithmetic-shift -100000000000000000000000000000000 -100)
    expected -79 but got -78
FAIL: (first-set-bit (expt -2 99))
    expected 99 but got 60
```

**Comment by ashinn** [comment]
Thanks for the report!  What platform is this?
The migration to (srfi 142) is still in-progress, and it's known some tests fail on Linux only, so I'm guessing this is the same problem.

**Comment by johnwcowan** [comment]
Linux and Cygwin.

**Comment by ashinn** [comment]
Actually, it was a 32-bit vs 64-bit problem.  I recall fixing similar cases for bitwise-and a while back and forgot to review and add tests for ior and xor.

This should be fixed as of commit 9cf8a3ddf3a83fb0985a082f1b589fe545ac9e64.

**Comment by johnwcowan** [comment]
That disposed of some but not all of the problems (on 32-bit Cygwin):

```
546 out of 558 (97.8%) tests passed in 0.166078090667725 seconds.
12 failures (2.2%).
FAIL: (copy-bit 64 340282366920938463463374607431768211456 #f)
    expected 340282366920938463463374607431768211456 but got 0
    on line 216 of file "chibi-test.scm"
FAIL: (copy-bit 64 -340282366920938463463374607431768211456 #f)
    expected -340282366920938463463374607431768211456 but got 0
    on line 218 of file "chibi-test.scm"
FAIL: (copy-bit 256 -340282366920938463463374607431768211456 #t)
    expected -340282366920938463463374607431768211456 but got 2135987035920910082279229616932235919179133537346503360456781002605296415371930057602824792965120
    on line 220 of file "chibi-test.scm"
FAIL: (bit-field-rotate 340282366920938463463374607431768211456 128 0...
    expected 340282366920938463463374607431768211456 but got 0
    on line 238 of file "chibi-test.scm"
FAIL: (bit-field-rotate 340282366920938463463374607431768211457 3 0 64)
    expected 340282366920938463463374607431768211464 but got 8
    on line 240 of file "chibi-test.scm"
FAIL: (bit-field-rotate 340282366920938463463374607431768211457 -3 0 64)
    expected 340282366920938463465680450440981905408 but got 2305843009213693952
    on line 242 of file "chibi-test.scm"
FAIL: (bit-field-reverse -2 0 31)
    expected -1073741825 but got 18446744072635809791
    on line 260 of file "chibi-test.scm"
FAIL: (bit-field-reverse -2 0 32)
    expected -2147483649 but got 18446744071562067967
    on line 262 of file "chibi-test.scm"
FAIL: (arithmetic-shift -100000000000000000000000000000000 -100)
    expected -79 but got -78
    on line 368 of file "chibi-test.scm"
FAIL: (arithmetic-shift -100000000000000000000000000000000 -100)
    expected -79 but got -78
    on line 372 of file "chibi-test.scm"
FAIL: (arithmetic-shift -100000000000000000000000000000000 -100)
    expected -79 but got -78
    on line 707 of file "chibi-test.scm"
FAIL: (arithmetic-shift -100000000000000000000000000000000 -100)
    expected -79 but got -78
    on line 711 of file "chibi-test.scm"
```
